llvm/test/CodeGen/RISCV/align.ll

```diff
@@ -2,8 +2,6 @@
 ; RUN:   | FileCheck %s -check-prefix=RV32I
 ; RUN: llc -mtriple=riscv32 -mattr=+c -verify-machineinstrs < %s \
 ; RUN:   | FileCheck %s -check-prefix=RV32C
-; RUN: llc -mtriple=riscv32 -mattr=+pref-func-align-32 -verify-machineinstrs < %s \
```

Comment:
Are there any processors that have non-default alignment? We could pass -mcpu and check for the `.p2align N` line so that we have test coverage.

Reply:
I don't know one. Maybe you can provide more information about SiFive processors in upstream?

Reply:
We haven't noticed any performance issues from instruction alignment on SiFive7 and are relying on linker to get this right. I'm not so sure what else we can do to have test coverage for this without an upstream processor that uses this TuneInfo 🤷

Reply:
I've just learned that different alignment values have different results on different workloads for us. Therefore, I think we should have way to control this value at compile time.

Reply:
There are IR attributes that override the CPU defaults. falign-loops and falign-functions uses those attributes.

Reply:
I found some descriptions in https://starfivetech.com/uploads/u74mc_core_complex_manual_21G1.pdf:
> 3.2.6 Instruction Fetch Unit
> The S7 instruction fetch unit is responsible for keeping the pipeline fed with instructions from
memory. The instruction fetch unit delivers up to 8 bytes of instructions per clock cycle to sup-
port superscalar instruction execution. _Fetches are always **word-aligned** and there is a one-
cycle penalty for branching to a 32-bit instruction that is not word-aligned._

It seems we'd better to align functions and hot loops to 4-byte to get best performance. But `C` extension is supported so the alignment would be 2-byte.
```
  // Function alignments.
  const Align FunctionAlignment(Subtarget.hasStdExtCOrZca() ? 2 : 4);
  setMinFunctionAlignment(FunctionAlignment);
  // Set preferred alignments.
  setPrefFunctionAlignment(Subtarget.getPrefFunctionAlignment());
  setPrefLoopAlignment(Subtarget.getPrefLoopAlignment());
```